lib/email/sender.rb

```diff
@@ -163,8 +163,11 @@ def send
 
         # WARNING: RFC claims you can not set the Return Path header, this is 100% correct
         # however Rails has special handling for this header and ends up using this value
-        # as the Envelope From address so stuff works as expected
-        @message.header[:return_path] = SiteSetting.reply_by_email_address.sub("%{reply_key}", "verp-#{email_log.bounce_key}")
+        # as the Envelope From address. Ideally it is supposed to be set by the mail server and the message can get flagged
+        # by antispam filters if the From address domain doesn't match the senders address domain due to DKIM and SPF failures
+        if SiteSetting.reply_by_email_address.split("@").last == @message.from&.split("@").last
```

Comment:
Hmm, can `@message.from` be `nil` at this point? If so, it will break because it should be `&.last`.
But I think `from` might have a different value than what you expect at this point. I assume it will look something like this: `Some Awesome Discourse Community <awesome@example.com>`

The condition wouldn't work in that case. So, yeah, this really needs some tests.

Reply:
Hmm okay. From what I understood the `from` should be set by the time this code executes and it should contain the senders email address (and it shouldn't be empty).

Is my understanding incorrect?

You're correct, it should be `@message.from&.split("@")&.last` just incase it's nil.